### ISSUE TITLE
slowness issue

### ISSUE THREAD

**Issue by flammie** [divvun/divvunspell#32]
I was demoing the morphology-aware spelling prediction stuff here at the conference and ran into a weird slowness issue I might need help to debug and profile, I have a zhfst and input that works instantaneously in hfst-ospell but takes 20 minutes in divvunspell, I thought first it was in my code, but it seems to happen in without analysis option as well:

```console
$ echo niwa | time hfst-ospell -a crk-neu.zhfst -n 10
...
1.13user 0.02system 0:01.18elapsed 97%CPU (0avgtext+0avgdata 54824maxresident)k
$ echo niwa | time cargo run -- suggest --archive crk-neu.zhfst -S -n 10
...
1146.24user 1.33system 19:08.48elapsed 99%CPU (0avgtext+0avgdata 126788maxresident)k
```

Perhaps I am doing something simple wrong? I can upload the zhfst after the conference with better network and battery

**Comment by bbqsrc** [comment]
Can help debug when you're back next week. Pathological cases are fuuuun.

**Comment by flammie** [comment]
Here's the file, renamed as github is being silly: [crk-neu.zhfst.zip](https://github.com/divvun/divvunspell/files/8796231/crk-neu.zhfst.zip). I'll be busy today but free most of the rest of the week.

**Comment by aarppe** [comment]
The *.hfst versions of morpheme completion FST can be slow, but the *.hfstol are decently faster.

**Comment by aarppe** [comment]
Hey guys - we'd be hoping to demo something about this, ideally a version running on a mobile device such as my iPad, next week at the Partnership General Meeting.

**Comment by flammie** [comment]
I was doing debugging / tracing and it seems that the system legit goes through a whole large search space but maybe c++ version has hardcoded timeout just. 

I might be off the track but the current code seems to sort and prune the suggestion list every new suggestion found which is probably necessary for keeping the suggestion list data structures in decent sizes and the weight limit that up-to-date but if I pull that out of inner loop I get some same results faster for this experiment? https://github.com/divvun/divvunspell/commit/e6e198bef07936636186ca2758ef0ada92ef5479

**Comment by bbqsrc** [comment]
@flammie yes, it's for weight limit and memory weight. When you say faster, how _much_ faster?

**Comment by flammie** [comment]
> 



> Uh, have you been running the debug version of divvunspell? You need to add `--release` to cargo run for optimisations to be enabled.

ach, does it have such huge effect too? It seems I need to recompile the world with --release now D:

if the release version handles this zhfst and input just fine then it's just that we'd like to have the attached zhfst perhaps in some dev beta keyboard for demoing...

**Comment by bbqsrc** [comment]
Yeah if it's just that we can do that.

**Comment by flammie** [comment]
Right, this is the release:
```
285.75user 2.38system 4:49.46elapsed 99%CPU (0avgtext+0avgdata 123276maxresident)k
0inputs+0outputs (0major+1444099minor)pagefaults 0swaps
```
it's down to 5 minutes from 20 so quarter time but probably no very useful for spelling corrections yet...

**Comment by bbqsrc** [comment]
Ooh, 4x faster. Cool, so we have some work to do here.

**Comment by aarppe** [comment]
Might the next step be to explore whether moving the weights as much to the beginning of the prefix-weighting FST, that is used as the error-model in the *.zhfst setup, would have some positive impact?

**Comment by aarppe** [comment]
And I'm wondering about other ways to split the weighted morpheme completion FST. Could there be a point in setting the spell-relax component as the error model (without weights), and the rest as the acceptor (with weights)?

**Comment by snomos** [comment]
@aarppe it would be better to keep the two discussions (slowness and morpheme completion) as separate threads. Could you create a new issue, and move your comments related to morpheme completion there? You can of course reference this issue in the new one :)
